File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
n, m = map(int, input().split())
arr = list(map(int, input().split()))
arr.sort()
start = 1
end = max(arr)

while start <= end:
    mid = (start + end) // 2
    tree = 0
    for i in arr:
        if i >= mid:
            tree += i - mid

    if tree >= m:
        start = mid + 1
    else:
        end = mid - 1

print(end)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const [N, height] = input.shift().split(" ").map(Number);
const tree = input[0]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);

let start = 0;
let end = tree[tree.length - 1];
let answer = Number.MIN_SAFE_INTEGER;

while(start <= end) {
    let mid = Math.floor((start + end) / 2)
    let sum = 0;
    for(let x of tree) {
        if(x > mid) sum += x-mid;
    }

    if(sum >= height) {
        if(mid > answer) answer = mid;
        start = mid+1;
    } else {
        end = mid - 1;
    }
}
console.log(answer);
